lib/puppet/parser/resource.rb

```diff
@@ -74,15 +74,12 @@ def add_edge_to_stage
   def evaluate
     return if evaluated?
     @evaluated = true
-    if klass = resource_type and ! builtin_type?
-      finish
-      evaluated_code = klass.evaluate_code(self)
```

Comment:
The `evalutate_code` is called from `validate` which is called from `finish`. The flow is correct but I admit that it looks rather odd. I'll change this.